projects/idea-hub-smart-contract-solidity/contracts/IdeaHub.sol

```diff
@@ -0,0 +1,288 @@
+        require(content.exists);
+        require(!content.isValidator[validator]);
+        content.isValidator[validator] = true;
+        content.validators.push(validator);
+        emit Validated(ideaNo, revisionNo, content.validators.length);
+    }
+
+
+    //// views ////
+
+    function ideas(uint256 ideaNo) public view override(IIdeaHub) returns (bytes32 id, bytes32 author, bool isPublished, uint256 revisionCount){
+        require(ideaNo > 0 && ideaNo <= ideaIds.length, "Index out of bounds");
+        bytes32 ideaId = ideaIds[ideaNo - 1];
+        Idea storage idea = _ideas[ideaId];
+        return (ideaId, users[idea.author].id, idea.published, idea.revisionList.length);
+    }
+
+    function ideaRevision(uint256 ideaNo, uint256 revisionNo) public view override(IIdeaHub) returns (bytes32 contentHash, uint256 validations){
+        require(ideaNo > 0 && ideaNo <= ideaIds.length, "Index out of bounds");
+        Idea storage idea = _ideas[ideaIds[ideaNo - 1]];
+        require(revisionNo > 0 && revisionNo <= idea.revisionList.length, "Index out of bounds");
+        bytes32 _contentHash = idea.revisionList[revisionNo - 1];
+        IdeaContent storage content = idea.revisions[_contentHash];
+        return (_contentHash, content.validators.length);
+    }
+
+    function ideaValidator(uint256 ideaNo, uint256 revisionNo, uint256 validatorNo) public view override(IIdeaHub) returns (address){
+        require(ideaNo > 0 && ideaNo <= ideaIds.length, "Index out of bounds");
+        Idea storage idea = _ideas[ideaIds[ideaNo - 1]];
+        require(revisionNo > 0 && revisionNo <= idea.revisionList.length, "Index out of bounds");
+        IdeaContent storage content = idea.revisions[idea.revisionList[revisionNo - 1]];
+        require(validatorNo > 0 && validatorNo <= content.validators.length, "Index out of bounds");
+        return content.validators[validatorNo - 1];
+    }
+
+    function authorId(address _address) public view override(IIdeaHub) returns (bytes32){
+        return _userId(_address, 1);
+    }
+
+    function authorAddress(bytes32 id) public view override(IIdeaHub) returns (address){
+        return _userAddress(id, 1);
+    }
+
+    function validatorId(address _address) public view override(IIdeaHub) returns (bytes32){
+        return _userId(_address, 2);
+    }
+
+    function validatorAddress(bytes32 id) public view override(IIdeaHub) returns (address){
+        return _userAddress(id, 2);
+    }
+
+    ///// Transaction functions /////
+
+    function registerAuthor(bytes32 userId, address _address, bytes32 r, bytes32 s, uint8 v) public override(IIdeaHub) nonZeroBytes(userId) {
+        AuthorRegistration memory registration = AuthorRegistration(userId, _address);
+        bytes32 message = hash(domainSeparator, hash(registration));
+        address signer = ecrecover(message, v < 27 ? v + 27 : v, r, s);
+        require(signer == _address);
+        _registerUser(userId, _address, 1);
```

Comment:
Please use constants instead of numbers 1 or 2 for roles here, would be more descriptive